 Weekly assignment 3

In this assignment we're going to merge two datafiles, clean them and explore the relations between the variables.

In [11]:
import pandas as pd
import seaborn as sns
from sklearn.preprocessing import normalize 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.neighbors import KNeighborsClassifier #Ik stop ze er gewoon allemaal meteen in
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

df = pd.read_csv("attrition.csv")
df

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0                    1          2  Life Sciences              1   
1                    8          1  Life Sciences              1   
2                    2          2          Other              1   
3                    3          4  Life Sciences              1   
4                    2          1        Medical              1   
...                ...        ...            ...            ...   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
0                  1  ...                         1            80   
1                  2  ...                         4            80   
2                  4  ...                         2            80   
3                  5  ...                         3            80   
4                  7  ...                         4            80   
...              ...  ...                       ...           ...   
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                  1               6                  4   
1                  3              10                  7   
2                  3               0                  0   
3                  3               8                  7   
4                  3               2                  2   
...              ...             ...                ...   
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9    

#Ik heb gekozen voor de employee attritition data set omdat hij mij het meest interessant leek

#de rest lijken alsof ze oninteressant zullen zijn in vergelijking met employee attritition 

#en "predict attrition" klinkt cooler want ik het doet me denken aan attrition in een oorlog of zoiets, ook al gaat het maar over medewerkers

In [14]:
dummies = pd.get_dummies(df['Attrition'])
df = pd.concat([df, dummies], axis=1) 
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0  ...                 0                 8                      0   
1  ...                 1                10                      3   
2  ...                 0                 7                      3   
3  ...                 0                 8                      3   
4  ...                 1                 6                      3   

   WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                1               6                  4   
1                3              10                  7   
2                3               0                  0   
3                3               8                  7   
4                3               2                  2   

   YearsSinceLastPromotion YearsWithCurrManager  No  Yes  
0                        0                    5   0    1  
1                        1                    7   1    0  
2                        0                    0   0    1  
3                        3                    0   1    0  
4                        2                    2   1    0  

[5 rows x 37 columns]

#Ik zorg er meteen voor dat attrition numeriek kan worden weergegeven 

In [15]:
X = df[['YearsAtCompany', 'TotalWorkingYears', 'DistanceFromHome', 'Education', 'TrainingTimesLastYear', 'WorkLifeBalance']]
y = df['Yes']
X

YearsAtCompany  TotalWorkingYears  DistanceFromHome  Education  \
0                  6                  8                 1          2   
1                 10                 10                 8          1   
2                  0                  7                 2          2   
3                  8                  8                 3          4   
4                  2                  6                 2          1   
...              ...                ...               ...        ...   
1465               5                 17                23          2   
1466               7                  9                 6          1   
1467               6                  6                 4          3   
1468               9                 17                 2          3   
1469               4                  6                 8          3   

      TrainingTimesLastYear  WorkLifeBalance  
0                         0                1  
1                         3                3  
2                         3                3  
3                         3                3  
4                         3                3  
...                     ...              ...  
1465                      3                3  
1466                      5                3  
1467                      0                3  
1468                      3                2  
1469                      3                4  

[1470 rows x 6 columns]

#Ik kies deze 7 variabelen voor de volgende redenen
#Yes: Het is interessanter om te weten welke mensen wel attritrion hebben
#YearsAtCompany: hoe lang ze bij het bedrijf werken kan misschien die attritrion veroorzaken
#TotalWorkingYears: hetzelfde als hierboven
#DistanceFromHome: hoe lang iemand moet reizen is altijd heel belangrijk
#Education: heeft hun niveau van educatie misschien invloed
#TrainingTimesLastYear: Ik wil zien of meer training dingen verbeteren of niet
#WorklifeBalance: Te veel werken is heel vaak een rede voor burn out 

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)
X_train.head() #ik split de training en de test

YearsAtCompany  TotalWorkingYears  DistanceFromHome  Education  \
99                 3                 17                23          3   
785               11                 14                20          4   
918               29                 31                 9          3   
1335               2                  7                19          4   
1182               3                  4                 1          4   

      TrainingTimesLastYear  WorkLifeBalance  
99                        3                4  
785                       6                3  
918                       5                2  
1335                      2                3  
1182                      6                3

In [17]:
knn = KNeighborsClassifier(5) 
knn = knn.fit(X_train, y_train) 
knn

KNeighborsClassifier()

In [18]:
knn.score(X_test, y_test)

0.7959183673469388

#hier volg ik eigenlijk gewoon het proces dat we eerder hebben gevolgd zodat ik de knn score kan zien

#Door het aantal mensen die wel attrition hebben als Y te nemen kan ik de invloed van die variabelen op het aantal attritrion berekenen 

#Ik zie hier dus dat het ongeveer 80% accuraat is als het gaat om berekenen van attrition 

In [19]:
y_test_pred = knn.predict(X_test) 
cm = confusion_matrix(y_test, y_test_pred) 
cm

array([[342,  22],
       [ 68,   9]])

In [20]:
conf_matrix = pd.DataFrame(cm, index=['Yes attrition (actual)', 'no attrition (actual)'], columns = ['Yes attrition (predicted)', 'No attrition (predicted)']) 
conf_matrix

Yes attrition (predicted)  No attrition (predicted)
Yes attrition (actual)                        342                        22
no attrition (actual)                          68                         9

#Wat betekent dit dus precies? Ik heb geen idee, daarom ga ik nu vragen aan mijn docent wat het betekent

#Wat ik wel weet is dat het model nu wel al gesplit is en dat hij getraind wordt

In [22]:
(342 + 9)/(68+342+22+9)

0.7959183673469388

#Nu weet ik dus dat mijn vorige berekening accuraat is omdat ik op hetzelfde uitkom

In [5]:
342 / (342 + 68) #precision

0.8341463414634146

#ik realiseer me dat ik de aantallen in de verkeerde vakjes heb gestopt, maar de aantallen zijn nog steeds correct

#oftewel, 83% van de predicted attrition is attrition 

In [8]:
(342) / (342 + 22) #recall

0.9395604395604396

94% van de echte attrition is berekend als attrition 

Zoals te zien is berekent het algoritme accuracy, precission en recall allemaal met 80% en erboven. Hieruit dat deze algoritme het berekenen van attrition ongeveer 80% accuraat is bij het berekenen van attrition en toekomstige attrition. Attrition in een bedrijf is erg serieus en kan een groot probleem veroorzaken als het niet correct kan worden voorspeld. Daarom concludeer ik dat 80% nog niet goed genoeg is voor een bedrijf. De recall staat wel op een acceptabele percentage.